src/connection_string.ts

```diff
@@ -785,26 +775,792 @@ function parseURI(uri: string): { srv: boolean; url: URL; hosts: string[] } {
+    password: typeof url.password === 'string' ? decodeURIComponent(url.password) : undefined
+  });
+
+  const urlOptions = new Map();
+  for (const key of url.searchParams.keys()) {
+    const loweredKey = key.toLowerCase();
+    const values = [...url.searchParams.getAll(key)];
+
+    if (values.includes('')) {
+      throw new MongoParseError('URI cannot contain options with no value');
+    }
+
+    if (urlOptions.has(loweredKey)) {
+      urlOptions.get(loweredKey)?.push(...values);
+    } else {
+      urlOptions.set(loweredKey, values);
+    }
+  }
+
+  const objectOptions = new Map(
+    Object.entries(options).map(([k, v]) => [k.toLowerCase(), v] as [string, any])
+  );
+
+  const allOptions = new Map();
+
+  const allKeys = new Set([
+    ...urlOptions.keys(),
+    ...objectOptions.keys(),
+    ...defaultOptions.keys()
+  ]);
+
+  for (const key of allKeys) {
+    const values = [];
+    if (urlOptions.has(key)) {
+      values.push(...urlOptions.get(key));
+    }
+    if (objectOptions.has(key)) {
+      values.push(objectOptions.get(key));
+    }
+    if (defaultOptions.has(key)) {
+      values.push(defaultOptions.get(key));
+    }
+    allOptions.set(key, values);
+  }
+
+  for (const [loweredKey, values] of allOptions.entries()) {
+    setOption(mongoOptions, loweredKey, values);
+  }
+
+  mongoOptions.credentials?.validate();
+  checkTLSOptions(mongoOptions);
+
+  return Object.freeze(mongoOptions) as Readonly<MongoOptions>;
+}
+
+function setOption(mongoOptions: any, loweredKey: string, values: unknown[]) {
+  const descriptor = descriptorFor(loweredKey);
+  const {
+    descriptor: { target, type, transform, deprecated },
+    key
+  } = descriptor;
+  const name = target ?? key;
+
+  if (deprecated) {
+    console.warn(`${key} is a deprecated option`);
+  }
+
+  switch (type) {
+    case 'boolean':
+      mongoOptions[name] = getBoolean(name, values[0]);
+      break;
+    case 'int':
+      mongoOptions[name] = getInt(name, values[0]);
+      break;
+    case 'uint':
+      mongoOptions[name] = getUint(name, values[0]);
+      break;
+    case 'string':
+      if (values[0] === undefined) {
+        break;
+      }
+      mongoOptions[name] = String(values[0]);
+      break;
+    case 'record':
+      if (!isRecord(values[0])) {
+        throw new TypeError(`${name} must be an object`);
+      }
+      mongoOptions[name] = values[0];
+      break;
+    case 'any':
+      mongoOptions[name] = values[0];
+      break;
+    default: {
+      if (!transform) {
+        throw new MongoParseError('Descriptors missing a type must define a transform');
+      }
+      const transformValue = transform({ name, options: mongoOptions, values });
+      mongoOptions[name] = transformValue;
+      break;
+    }
+  }
+}
+
+function toHostArray(hostString: string) {
+  const parsedHost = new URL(`mongodb://${hostString.split(' ').join('%20')}`);
+
+  let socketPath;
+  if (hostString.match(/\.sock/)) {
```

Comment:
```suggestion
  if (hostString.match(/^(rel)?\/.+\.sock$/)) {
```

Reply:
What's the "rel" bit matching for?

Reply:
I thought it was required for the [relative path domain sockets](https://github.com/mongodb/specifications/blob/master/source/connection-string/tests/valid-unix_socket-relative.yml) but it actually looks optional. I've updated it to something that should work.

Here's a little script I wrote to test it:
```
const abs = require('./test/spec/connection-string/valid-unix_socket-absolute.json');
const rel = require('./test/spec/connection-string/valid-unix_socket-relative.json');
const DOMAIN_SOCKET_REGEX = /mongodb-\d+\.sock$/;
abs.tests.concat(rel.tests).forEach(test =>
  test.hosts.forEach(host => {
    if (host.type === 'unix') {
      if (!host.host.match(DOMAIN_SOCKET_REGEX)) {
        console.log(`Regex doesnt match ${host.host}`);
      }
    }
  })
);
```

Reply:
Yea I think the `rel` part is a stand in for any relative directory name, its just making sure we don't depend on a leading slash seems like.
Does raise the question tho, a socket can be named anything and doesn't have to have the .sock at the end, I wonder if there's another method drivers use to determine unix socket hosts. I took a look at go, they determine based on a "sock" suffix, so no dot, and has to be at the end, I guess this is just they way of things.

Reply:
> Does raise the question tho, a socket can be named anything and doesn't have to have the .sock at the end,

True in general, but `mongod` doesn't let you specify a custom name for it, just a custom path. It's always named `mongodb-${port}.sock`. That said, I guess people could be using a symlink named something else entirely.

Reply:
Ah I just assumed it was customizable, then should we actually relax it to just a suffix of `sock` or the more strict `/mongodb-\d+\.sock$/`.  I'm indifferent, but yea it could work in reverse, valid hostnames can end in `.sock` although 🤞 hopefully not

Reply:
Looks like the [python driver](https://github.com/mongodb/mongo-python-driver/blob/d1fd3f7e982f5d1ae3e8b8b8edeba7e146bb02f8/pymongo/pool.py#L947) just checks that it ends with `.sock`, we might as well stick with that as a convention since there's no specification around this.